### PR TITLE
SCLORG sanity checks over all repositories

### PR DESCRIPTION
**Pull request by phracek**
The aim of this daily checker is:

* detect if .exclude and Dockerfile are present and think about usage .devel repo because of version is not tested
* detect if .exclude file is present and Dockerfile not
* detect if .devel_repo is present and GA is not reached.

What is the outcome.
* In case of script does not detect any issue. Relevant users will receive an email like:
```bash
SCLORG: sanity checker does not hit any issues in https://github/sclorg containers
```
* Incase of script hit some issue. Relevant users will receive an email like:
```bash
SUBJECT: SCLORG: sanity checker hit some issues in https://github/sclorg containers
BODY: SCLORG sanity checker hit any issues for one of these repositories:
s2i-perl-container
s2i-nodejs-container
s2i-php-container
s2i-ruby-container
mariadb-container
nginx-container

Attachment for s2i-perl-container will look like
--- 5.34:
File .exclude-c9s is present in and Dockerfile.c9s as well. The version 5.34 is not tested.
Please think about usage '.devel-repo-c9s' or check if image does not reach EOL.

```
Follow up pull request will be also checking GA of images which will be deprecated soon.
Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

### CODE REVIEW

**Comment by zmiklank** [comment]
Sorry, I am missing some context. What is this "daily checker" please?

File: daily_tests/sclorg_sanity_tests.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        self.report_file: str = ""
+        self.report_text_filename: str = ""
+
+    def write_to_textfile(self, msg):
+        print(msg)
+        with open(self.report_file, "a") as fw:
+            fw.write(msg)
+            fw.write("\n")
+
+    def create_tmp_dir(self):
```
Comment (hhorak) [inline daily_tests/sclorg_sanity_tests.py:66]: I don't see the tmpdir is cleaned -- is it staying created on purpose? There's been effort cleaning after images' tests, so maybe this test might clean after itself as well, alternatively `--no-clean` option can be added if there is a reason to keep the working dirs for further analysis.

File: daily_tests/sclorg_sanity_tests.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+from email.mime.application import MIMEApplication
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from tempfile import TemporaryDirectory
+from pathlib import Path
+from typing import Dict
+from shutil import rmtree
+from git import Repo
+from git.exc import GitCommandError
```
Comment (hhorak) [inline daily_tests/sclorg_sanity_tests.py:17]: Some of the modules are likely not in the default set of components -- shouldn't we have either one common requirements.txt, or better include all scripts into a separate directory (with proper requirements.txt and README.md for the single tool) and then only symlink in the daily_tests dir? (just an idea)

File: daily_tests/sclorg_sanity_tests.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                attach = MIMEApplication(open(log, 'r').read(), Name=os.path.basename(str(log)))
+                attach.add_header('Content-Disposition', 'attachment; filename="{}"'.format(os.path.basename(str(log))))
+                msg.attach(attach)
+        smtp = smtplib.SMTP("127.0.0.1")
+        smtp.sendmail(send_from, send_to, msg.as_string())
+        smtp.close()
+
+
+if __name__ == "__main__":
+    sanity_checker = SclOrgSanityChecker()
```
Comment (hhorak) [inline daily_tests/sclorg_sanity_tests.py:224]: We should have some comment that explains what the tool does. Alternatively, even when the tool does not accept options at this point, we can use argparese and have standardized `-h`, `--help` with a reasonable explanation. (just an idea, not a requirement to implement it this way)

File: daily_tests/sclorg_sanity_tests.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            self.report_file = self.tmp_path_dir / f"{repo}.log"
+
+            os.chdir(self.tmp_path_dir / repo)
+            versions_to_check = self.get_all_supported_versions(repo_name=repo)
+
+            for ver in versions_to_check:
+                sanity_ok = True
+                self.write_to_textfile(f"--- {ver}:")
+                if not self.check_exclude_file(ver):
+                    sanity_ok = False
```
Comment (hhorak) [inline daily_tests/sclorg_sanity_tests.py:180]: Do I understand correctly it only prints list of repos where an issue was detected? Maybe we can report what kind of issue right away, so it's more clear from the mail...

**Comment by hhorak** [comment]
Thanks for the initial work, this is gonna be useful. Several comments in-line.

File: daily_tests/sclorg_sanity_tests.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+from tempfile import TemporaryDirectory
+from pathlib import Path
+from typing import Dict
+from shutil import rmtree
+from git import Repo
+from git.exc import GitCommandError
+
+SCLORG_REPOS = [
+    # Format is 'repo_name' and 'package_name'
+    ("s2i-base-container", "s2i-base"),
```
Comment (zmiklank) [inline daily_tests/sclorg_sanity_tests.py:21]: I there any reason why the base container has the `s2i` prefix when others do not have it?
```suggestion
    ("s2i-base-container", "base"),
```

File: daily_tests/sclorg_sanity_tests.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+                    f"For version {ver} .exclude-{os_ver} is present in {ver}"
+                    f"but Dockerfile.{os_ver} not."
+                )
+                self.write_to_textfile(f"Think about for removal .exclude-{os_ver}.")
+                ret_val = False
+        return ret_val
+
+    def check_devel_repo_file(self, ver: str):
+        ret_val = True
+        print(f"Check .exclude file for {ver}.")
```
Comment (zmiklank) [inline daily_tests/sclorg_sanity_tests.py:97]:
```suggestion
        print(f"Check .devel_repo file for {ver}.")
```

File: daily_tests/sclorg_sanity_tests.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        for os_ver in OS_HOSTS:
+            devel_repo_file = self.tmp_path_repo / ver / f".devel-repo-{os_ver}"
+            dockerfile = self.tmp_path_repo / ver / f"Dockerfile.{os_ver}"
```
Comment (zmiklank) [inline daily_tests/sclorg_sanity_tests.py:100]: I see a lot of duplicity in `check_devel_repo_file` and ` check_exclude_file`. Could not it be done something with it?

**Review by zmiklank — CHANGES_REQUESTED** [review]
Thank you for working on the checker.
Please see my comments. Also, would it be possible to make the commit message more verbose, so we will know in the future what the commit does?

Another thing: where will this script run? Would it be triggered periodically via GitHub Actions? Or do you have something else in mind?

File: daily_tests/sclorg_sanity_tests.py (hunk 2 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:17]: Make sense to me. I will drive it.

File: daily_tests/sclorg_sanity_tests.py (hunk 3 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:224]: Good idea.

File: daily_tests/sclorg_sanity_tests.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    "centos7",
+    "c9s",
+    "c8s",
+    "rhel7",
+    "rhel8",
+    "rhel9"
+]
+
+# This is reference link to internal repository
+URL = "https://url.corp.redhat.com/application-streams-yaml"
```
[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:47]: The url contains information about lifestream EOL and GA information.

File: daily_tests/sclorg_sanity_tests.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    def send_email(self, failed_repos, result: bool = False, ):
+        if result:
+            subject_msg = "SCLORG: sanity checker did not hit any issue in https://github/sclorg containers"
+            message = f"SCLORG sanity checker did not hit any issues for {SCLORG_REPOS}"
+        else:
+            subject_msg = "SCLORG: sanity checker hit some issues in https://github/sclorg containers"
+            message = f"SCLORG sanity checker hit any issues for one of these repositories:\n"
+            message += "\n".join(failed_repos)
+        send_from = "phracek@redhat.com"
+        send_to = "phracek@redhat.com"
```
[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:205]: It is planned to have all containerization and some Apps folks. I will update it.

File: daily_tests/sclorg_sanity_tests.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+    def check_tested_version(self, ver: str, pkg_name: str):
+        ret_val = True
+        print(f"Check .exclude and Dockerfile for ver {ver}")
+        from datetime import date
+        today = date.today()
+        ymd = today.strftime("%Y%m%d")
+        for os_ver in OS_HOSTS:
+            exclude_file = self.tmp_path_repo / ver / f".exclude-{os_ver}"
+            dockerfile = self.tmp_path_repo / ver / f"Dockerfile.{os_ver}"
```
[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: this is fixed by commit https://github.com/sclorg/ci-scripts/pull/75/commits/a4ad5e6019012baed5db30ff9b7f128a4bff869f

File: daily_tests/sclorg_sanity_tests.py (hunk 7 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: THis is fixed by commit https://github.com/sclorg/ci-scripts/pull/75/commits/a4ad5e6019012baed5db30ff9b7f128a4bff869f

File: daily_tests/sclorg_sanity_tests.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: This is fixed by commit https://github.com/sclorg/ci-scripts/pull/75/commits/a4ad5e6019012baed5db30ff9b7f128a4bff869f

File: daily_tests/sclorg_sanity_tests.py (hunk 3 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:265]: Fixed by https://github.com/sclorg/ci-scripts/pull/75/commits/a4ad5e6019012baed5db30ff9b7f128a4bff869f

File: daily_tests/sclorg_sanity_tests.py (hunk 2 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:18]: Fixed by https://github.com/sclorg/ci-scripts/pull/75/commits/35c358e11431045df368d373a2b9661819e11a77

File: daily_tests/sclorg_sanity_tests.py (hunk 4 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: It is done during the print like:

```bash
--- 5.34:
File .exclude-c9s is present in and Dockerfile.c9s as well. The version 5.34 is not tested.
Please think about usage '.devel-repo-c9s' or check if image does not reach EOL.
```

File: daily_tests/sclorg_sanity_tests.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: Fixed by commit https://github.com/sclorg/ci-scripts/pull/75/commits/c006f9027ed5aba80b0a30c065a179d5253a2234

File: daily_tests/sclorg_sanity_tests.py (hunk 1 above)
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:68]: It is fixed at the end of run_check function https://github.com/sclorg/ci-scripts/pull/75/files#diff-31589a09c0dea50a4a46f7ec84601ba88c0f9c5bb220755146cfbaafbc89d5c9R234

File: daily_tests/sclorg_sanity_tests.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+                        sanity_ok = False
+                    if not self.check_tested_version(ver, os_ver):
+                        sanity_ok = False
+                if sanity_ok:
+                    os.unlink(self.report_file)
+                else:
+                    if repo not in failed_repos:
+                        failed_repos.append(repo)
+                    global_result_flag = False
+        self.send_email(failed_repos=failed_repos, result=global_result_flag)
```
[the remark this answers is not in the document]
Comment (hhorak) [reply daily_tests/sclorg_sanity_tests.py:232]: I even think the sending itself could be optional -- sometimes one can run the test locally (even repeatedly) to check whether there are still some issues and in that case, sending mail is not desired.

[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py:232]: I will do it.

[the remark this answers is not in the document]
Comment (phracek) [reply daily_tests/sclorg_sanity_tests.py]: Fixed by https://github.com/sclorg/ci-scripts/pull/75/commits/62139112baf0989a8e4eaf819cace85589397706

File: daily_tests/sclorg_sanity_tests.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+    "rhel7",
+    "rhel8",
+    "rhel9"
+]
+
+# This is reference link to internal repository
+EOL_GA_URL = "https://url.corp.redhat.com/application-streams-yaml"
+
+
+class GitHubSanityChecker(object):
```
Comment (zmiklank) [inline daily_tests/sclorg_sanity_tests.py:50]: nickpick: I found the original name more fitting (SclOrgSanityChecker.)

**Review by zmiklank — APPROVED** [review]
One nitpick, esle LGTM.

**Comment by phracek** [comment]
Let's get merge it and deploy it. The first time it will be executed manually and the next run will be each Monday at 8 am.

**Pull request merged**